### PR TITLE
Adaptive attribute and parallelization implementation

### PR DESCRIPTION
**Pull request by VictorianHues**
Adaptive attributes implemented for env coefficient and peer pressure coefficient, but env portion reverted after deciding against its use. #17 

Basic parallelization added for grid space iterations. Could be removed and use parallel batch runs instead #25 

Plotting functions with similar functionality merged to allow for attribute inputs from the model instead of directly defining a different function for each attribute

Some messy data analysis was also implemented, not sure how useful all of it will be #26

### CODE REVIEW

File: scripts/base_model.py [hunk 1]
```diff
@@ -32,9 +39,9 @@ def env_status_fn():
 
     def peer_pressure_coeff_fn():
         if rng:
-            return rng.uniform(0, 1)
+            return rng.uniform(1.0, 1.0)
         else:
-            return np.random.uniform(0.5, 1.0)
+            return np.random.uniform(1.0, 1.0)
```
Comment (henry-zwart) [inline scripts/base_model.py:44]: Is the U(1,1) distribution intentional here?

File: scripts/base_model.py [hunk 2]
```diff
@@ -43,38 +50,92 @@ def env_perception_coeff_fn():
...
 
-    plot_current_agent_env_status(model)
-    plot_current_agent_actions(model)
+    plot_current_grid_state(
+        model.get_agent_attribute_at_time("agent_env_status_history", -1),
+        colormap="RdYlGn",
+        title="Agent Environment Status",
+        colorbar_label="Environment Status",
+        clim=(0, 1),
+    )
```
Comment (henry-zwart) [inline scripts/base_model.py:78]: This is a nice pattern :)

File: src/abm_project/agent.py [hunk 3]
```diff
@@ -20,13 +20,18 @@ class Agent:
     ACTIONS = [-1, 1]
     DEFAULT_MEMORY_COUNT = 1
     DEFAULT_ENV_UPDATE_OPTION = "linear"
+    DEFAULT_ADAPTIVE_ATTR_OPTION = None
+    DEFAULT_RATIONALITY = 0.5
```
Comment (henry-zwart) [inline src/abm_project/agent.py:24]: What's the reasoning for the default being 0.5? Perhaps we could scale the utility calculation by 1/2 prior to multiplying rationality so that the default rationality can be 1.

File: src/abm_project/agent.py [hunk 4]
```diff
@@ -70,6 +85,144 @@ def __init__(
...
+
+            consensus = max(proportion_action_pos, proportion_action_neg)
+            majority_action = (
+                1 if proportion_action_pos >= proportion_action_neg else -1
+            )
+
+            norm_consensus = (consensus - 0.5) * 2  # -1 to 1
+            confidence = norm_consensus  # 1 / (1 + np.exp(-k * norm_consensus))
+
+            confidence = 2 * (confidence - 0.5)  # [-1, 1]
```
Comment (henry-zwart) [inline src/abm_project/agent.py:169]: This value is in $[-3, 1]$ due to the rescaling on line 166. Perhaps this line is a duplicate?

File: src/abm_project/agent.py [hunk 5]
```diff
@@ -70,6 +85,144 @@ def __init__(
...
+            )
+
+            majority_flipped = majority_action != previous_majority
+            if majority_flipped:
+                confidence *= -1
+
+            if agreed_with_majority:
+                new_coeff = self.peer_pressure_coeff[-1] + learning_rate * confidence
+            else:
+                new_coeff = self.peer_pressure_coeff[-1] - learning_rate * confidence
```
Comment (henry-zwart) [inline src/abm_project/agent.py:190]: If I'm understanding this correctly, it's saying:
- If previously you agreed with your neighbors, but they have since changed their opinion, they now have less influence
- If previously you disagreed, but now agree, then they gain influence

Is that right?

File: src/abm_project/agent.py [hunk 6]
```diff
@@ -70,6 +85,144 @@ def __init__(
...
+            self.previous_majority = majority_action
+
+            # print(f"Peer Pressure Coeff Update: {self.peer_pressure_coeff[-1]} ->
+            # {new_coeff}, Action: {self.past_actions[-1]},
+            # Majority Action: {majority_action},
+            # Confidence: {confidence},
+            # Proportion Pos: {proportion_action_pos},
+            # Proportion Neg: {proportion_action_neg}")
+
+            new_coeff = np.clip(new_coeff, 0, 1)
```
Comment (henry-zwart) [inline src/abm_project/agent.py:203]: Duplicate line

File: src/abm_project/agent.py [hunk 7]
```diff
@@ -70,6 +85,144 @@ def __init__(
...
+        if len(self.env_perception_coeff) > self.memory_count:
+            self.env_perception_coeff.pop(0)
+
+    def update_peer_pressure_coeff(self, all_peer_actions: np.ndarray) -> float:
+        """Update the agent's peer pressure coefficient.
+
+        This coefficient is used to calculate the cost
+        of deviating from the average peer action.
+        """
+        if self.adaptive_attr_option == "bayesian":
```
Comment (henry-zwart) [inline src/abm_project/agent.py:150]: This adaptation method is interesting -- I'd like to see the mathematical form :)

File: src/abm_project/agent.py [hunk 8]
```diff
@@ -70,6 +85,144 @@ def __init__(
...
+            # print(f"Peer Pressure Coeff Update: {self.peer_pressure_coeff[-1]} ->
+            # {new_coeff}, Action: {self.past_actions[-1]},
+            # Majority Action: {majority_action},
+            # Confidence: {confidence},
+            # Proportion Pos: {proportion_action_pos},
+            # Proportion Neg: {proportion_action_neg}")
+
+            new_coeff = np.clip(new_coeff, 0, 1)
+        elif self.adaptive_attr_option == "logistic_regression":
+            k = 10
```
Comment (henry-zwart) [inline src/abm_project/agent.py:205]: $k$ should probably be a model-level attribute

File: src/abm_project/agent.py [hunk 9]
```diff
@@ -20,13 +20,18 @@ class Agent:
...
+    DEFAULT_RATIONALITY = 0.5
 
     def __init__(
         self,
         id: int,
         memory_count: int = DEFAULT_MEMORY_COUNT,
         rng: np.random.Generator = None,
         env_update_option: str = DEFAULT_ENV_UPDATE_OPTION,
+        adaptive_attr_option: str = DEFAULT_ADAPTIVE_ATTR_OPTION,
+        peer_pressure_learning_rate=0.2,
```
Comment (henry-zwart) [inline src/abm_project/agent.py:33]: I wonder what would happen if this was heterogeneous. Maybe not varying continuously, but one of two values for each agent. For instance, "stubborn in own opinions" vs. "follows the trend".

The Tilman paper discusses how differences in agents' preference for leading vs. following change can result in different types of dynamics (stable, collapse, cyclic).

File: src/abm_project/agent.py (hunk 4 above)
Comment (henry-zwart) [reply src/abm_project/agent.py:169]: Ah, nevermind! The consensus is at least 0.5 due to the max.

File: src/abm_project/agent.py [hunk 10]
```diff
@@ -70,6 +85,144 @@ def __init__(
...
+            agreement = 1 if self.past_actions[-1] == majority_action else -1
+
+            delta = learning_rate * logit_confidence * agreement
+            new_coeff = np.clip(self.peer_pressure_coeff[-1] + delta, 0, 1)
+        else:
+            new_coeff = self.peer_pressure_coeff[-1]
+
+        self.peer_pressure_coeff.append(new_coeff)
+        if len(self.peer_pressure_coeff) > self.memory_count:
+            self.peer_pressure_coeff.pop(0)
```
Comment (henry-zwart) [inline src/abm_project/agent.py:224]: Popping from the start is fairly expensive if happening often. Might be worth making this a ring buffer.

File: src/abm_project/batch_run_tools.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+        model_class (type): The class of the model to instantiate.
+        steps (int): Number of simulation steps to run for each model.
+        **kwargs: Additional keyword arguments to pass to the model constructor.
+
+    Returns:
+        list[BaseModel]: List of model instances after running the simulation.
+    """
+    models = []
+    for _ in range(num_runs):
+        rng = np.random.default_rng()
```
Comment (henry-zwart) [inline src/abm_project/batch_run_tools.py:26]: Perhaps give this a fixed random seed per model, e.g., the idx in the models array.

**Review by henry-zwart — APPROVED** [review]
Looks good. I've left a few comments. The batch sim script seems to run okay on my machine, though had a 7h time estimate so I only ran it for a couple of iterations.

I get a warning from SALib that saltelli sampling is deprecated and `salib.sample.sobol` should be used instead.

File: src/abm_project/agent.py (hunk 3 above)
Comment (VictorianHues) [reply src/abm_project/agent.py:24]: Default 1.0 would be more reasonable, definitely something to change
